storage: Transfer leases based on request load and latency

a-robinson opened the pull request:
i.e. "Leaseholders follow the sun"

Performance on my macbook on the four main workloads detailed in #13232, with all workloads involving 32 workers and 3 nodes, each in a different locality:

1. 0ms of latency between each node, round-robin balanced load

1a) Without this change (control group):
```
_elapsed__ops/sec__average__latency___errors_replicas_______________1:1_______________2:2_______________3:3
    1m0s   1899.0   2469.2   12.9ms        0      720         240/87/47          240/74/0          240/73/4
    2m0s   1542.7   2348.4   13.6ms        0     1305        435/149/59         435/139/2        435/142/10
    3m0s   1370.3   2287.4   14.0ms        0     1587        529/175/62         529/180/8        529/169/10
    4m0s   2257.4   2248.7   14.2ms        0     2439        813/283/63         813/263/8        813/261/21
    5m0s   1790.8   2188.9   14.6ms        0     3084       1028/338/71       1028/356/17       1028/327/21
```

1b) With this change
```
_elapsed__ops/sec__average__latency___errors_replicas_______________1:1_______________2:2_______________3:3
    1m0s   2135.6   2387.1   13.4ms        0      714        238/145/43          238/46/1          238/40/1
    2m0s   1915.7   2214.3   14.4ms        0     1227        409/185/92         409/112/3         409/105/1
    3m0s   2105.5   2194.5   14.6ms        0     1560       520/197/133         520/156/7         520/159/1
    4m0s   1767.6   2149.7   14.9ms        0     2250       750/256/156         750/244/7         750/239/1
    5m0s   1780.7   2109.3   15.2ms        0     3039      1013/342/161        1013/341/7        1013/323/1
```

2. 50ms of latency between each node, round-robin balanced load

2a) Without this change (control group):
```
_elapsed__ops/sec__average__latency___errors_replicas_______________1:1_______________2:2_______________3:3
    1m0s    851.2    829.0   38.6ms        0      213          71/26/15           71/19/0           71/20/2
    2m0s    905.8    840.0   38.1ms        0      405         135/46/21          135/40/0          135/43/2
    3m0s    855.3    843.2   37.9ms        0      786         262/90/24          262/80/0          262/84/2
    4m0s    877.4    845.5   37.8ms        0      798         266/89/24          266/82/0          266/86/2
    5m0s    875.9    846.4   37.8ms        0     1071        357/114/24         357/116/1         357/114/2
    6m0s    825.7    846.7   37.8ms        0     1545        515/175/26         515/162/1         515/171/2
```

2b) With this change
```
_elapsed__ops/sec__average__latency___errors_replicas_______________1:1_______________2:2_______________3:3
    1m0s    837.9    824.1   38.8ms        0      213          71/21/22           71/22/8           71/22/9
    2m0s    858.8    838.0   38.2ms        0      405         135/44/26         135/43/14         135/43/16
    3m0s    869.8    842.5   38.0ms        0      786         262/84/31         262/87/19         262/86/21
    4m0s    818.8    845.4   37.8ms        0      792         264/86/37         264/86/24         264/86/25
    5m0s    824.5    844.1   37.9ms        0     1077        359/118/39        359/117/28        359/116/26
```

3. 0ms of latency between each node, all workers hitting node 2

3a) Without this change (control group):
```
_elapsed__ops/sec__average__latency___errors_replicas_______________1:1_______________2:2_______________3:3
    1m0s   2375.7   2425.4   13.2ms        0      741        247/129 40          247/39 0          247/75 0
    2m0s   2383.6   2349.5   13.6ms        0     1413        471/180 86         471/138 0         471/147 8
    3m0s   2258.8   2221.4   14.4ms        0     1560        520/177 99         520/166 0        520/172 14
    4m0s   2118.4   2141.9   14.9ms        0     2261       754/257 100         754/239 0        753/248 15
    5m0s   1790.7   2077.2   15.4ms        0     3063      1021/343 105        1021/331 0       1021/341 15
    6m0s   2066.8   2017.6   15.9ms        0     3099      1033/343 105        1033/336 0       1033/348 15
```

3b) With this change
```
_elapsed__ops/sec__average__latency___errors_replicas_______________1:1_______________2:2_______________3:3
    1m0s   2020.9   2268.8   14.1ms        0      648        216/132/43          216/54/2          216/25/3
    2m0s   1917.6   2191.6   14.6ms        0     1224        408/173/93         408/145/4          408/84/7
    3m0s   2120.5   2157.2   14.8ms        0     1560       520/172/137        520/208/10         520/133/7
    4m0s   2357.5   2147.6   14.9ms        0     2319       773/226/161        773/316/14         773/224/7
    5m0s   1948.3   2101.6   15.2ms        0     3051      1017/296/161       1017/440/15        1017/275/7
```

4. 50ms of latency between each node, all workers hitting node 2

4a) Without this change (control group):
```
_elapsed__ops/sec__average__latency___errors_replicas_______________1:1_______________2:2_______________3:3
    1m0s    743.7    659.7   48.5ms        0      213     71/23     71/18     71/25
    2m0s    861.8    728.9   43.9ms        0      405    135/43    135/42    135/44
    3m0s    857.9    765.0   41.8ms        0      699    233/69    233/75    233/75
    4m0s    845.9    786.2   40.7ms        0      792    264/84    264/86    264/87
    5m0s    831.4    797.9   40.1ms        0      897    299/90    299/95   299/100
    6m0s    796.9    805.4   39.7ms        0     1488   496/157   496/168   496/162
```
4b) With this change
```
_elapsed__ops/sec__average__latency___errors_replicas_______________1:1_______________2:2_______________3:3
    1m0s   1809.7   1047.9   30.5ms        0      258          86/12/23           86/51/4           86/11/8
    2m0s   1937.6   1472.3   21.7ms        0      780         260/33/24         260/189/4         260/33/10
    3m0s   1963.9   1629.6   19.6ms        0     1335         445/65/24         445/305/7         445/61/10
    4m0s   1899.9   1690.4   18.9ms        0     1578         526/72/24        526/376/15         526/71/10
    5m0s   1692.7   1707.7   18.7ms        0     2241        747/118/25        747/518/25         747/95/10
    6m0s   1703.5   1709.7   18.7ms        0     2844        948/140/27        948/651/33        948/147/13
```

Other interesting experiments of note:

* Letting the server continue to run for a long time, the performance degrades quite a lot as time goes on. I saw similar degradation without the heuristic, so this may have just been my laptop slowing down as it got hotter and/or ran out of memory. This is with no added latency, imbalanced load, after this change:

```
_elapsed__ops/sec__average__latency___errors_replicas_______________1:1_______________2:2_______________3:3
    1m0s   2076.8   2563.1   12.5ms        0      774        258/165 43          258/67 1          258/22 2
    2m0s   2098.8   2439.2   13.1ms        0     1473        491/235 94         491/165 3          491/84 3
    3m0s   2069.8   2332.2   13.7ms        0     1560       520/211 142         520/187 9         520/119 3
    4m0s   2113.4   2278.4   14.0ms        0     2568       856/285 189        856/339 10         856/221 4
    5m0s   1177.3   2205.6   14.5ms        0     3093      1031/331 199       1031/420 12        1031/273 4
    6m0s   1936.5   2169.5   14.7ms        0     3099      1033/319 213       1033/416 22        1033/290 4
    7m0s   1914.6   2140.8   14.9ms        0     3351      1117/334 227       1117/454 23        1117/321 4
    8m0s   1876.8   2108.3   15.2ms        0     4347      1449/421 249       1449/603 29        1449/413 7
    9m0s   1839.7   2070.9   15.5ms        0     5304      1768/487 250       1768/733 37        1768/487 7
   10m0s   1678.5   2034.2   15.7ms        0     5970      1990/554 250       1990/858 46        1990/552 7
   11m0s   1672.1   2004.7   16.0ms        0     6153      2051/577 250       2051/877 47        2051/577 7
   12m0s   1802.8   1979.1   16.2ms        0     6186      2062/580 250       2062/890 47       2062/568 17
   13m0s   1501.5   1955.5   16.4ms        0     6186      2062/569 254       2062/888 52       2062/570 18
   14m0s   1677.1   1929.0   16.6ms        0     6198      2066/570 260       2066/897 53       2066/574 19
   15m0s   1240.4   1909.5   16.8ms        0     6396      2132/563 283       2132/933 56       2132/596 20
   16m0s   1150.8   1879.2   17.0ms        0     6819      2273/613 283       2273/958 66       2273/632 20
   17m0s   1433.8   1851.3   17.3ms        0     7417      2473/656 283      2471/1026 68       2473/684 20
   18m0s   1334.5   1828.9   17.5ms        0     8277      2759/737 283      2759/1170 69       2759/748 21
   19m0s   1410.5   1807.0   17.7ms        0     8973      2991/756 284      2991/1249 70       2991/767 23
   20m0s   1497.4   1786.2   17.9ms        0     9836      3278/813 290      3279/1386 73       3279/855 23
```

* Completely stopping and restarting a cluster with all the load focused on one node makes for imbalanced leases that never really get rebalanced, even though there's no latency. This is taking the run above and restarting it:

```
_elapsed__ops/sec__average__latency___errors_replicas_______________1:1_______________2:2_______________3:3
    1m0s   1507.8   1626.8   19.7ms        1    11073         3691/65 0      3691/3337 52         3691/59 0
    2m0s   1602.8   1583.8   20.2ms        1    11615         3872/86 0     3872/3308 107         3871/82 0
    3m0s   1621.7   1556.0   20.6ms        3    11982        3994/103 0     3994/3239 161        3994/100 0
    4m0s   1293.8   1540.7   20.8ms        4    12201        4067/140 0     4067/3234 216        4067/140 0
    5m0s   1307.9   1512.2   21.2ms        4    12303        4101/159 0     4101/3317 270        4101/165 0
```

* Running with `MinLeaseTransferStatsDuration` set to 1 minute rather than 10 seconds, with 50ms of latency and skewed load:

```
_elapsed__ops/sec__average__latency___errors_replicas_______________1:1_______________2:2_______________3:3
    1m0s    565.5    550.6   58.1ms        0      156           52/45/0            52/0/0            52/0/0
    2m0s    887.3    621.3   51.5ms        0      402         134/80/31          134/47/0           134/3/0
    3m0s   1385.7    800.8   40.0ms        0      735         245/89/67         245/131/0           245/8/0
    4m0s   1665.5    974.8   32.8ms        0      882        294/75/101         294/195/0          294/10/0
    5m0s   1752.4   1137.5   28.1ms        0     1545       515/112/114         515/371/2          515/24/0
    6m0s   1869.7   1243.0   25.7ms        0     1623       541/105/124         541/388/8          541/40/0
    7m0s   1815.7   1311.0   24.4ms        0     2397       799/159/136         799/555/9          799/72/0
    8m0s   1765.2   1367.5   23.4ms        0     2967       989/169/145        989/709/22         989/103/0
    9m0s   1619.8   1404.7   22.8ms        0     3102      1034/157/156       1034/751/26        1034/119/0
   10m1s   1681.2   1430.3   22.4ms        0     3153      1051/160/156       1051/755/30        1051/126/0
   11m0s   1733.8   1449.5   22.1ms        0     3810      1270/193/156       1270/899/31        1270/159/1
   12m0s   1643.2   1462.7   21.9ms        0     4605      1535/255/156      1535/1065/31        1535/198/1
   13m0s   1437.4   1464.7   21.8ms        0     5268      1756/287/156      1756/1220/32        1756/226/1
```

* Continuing the above (50ms latency, imbalanced load, minimum stats duration of 1 minute) after a full restart of the nodes:

```
_elapsed__ops/sec__average__latency___errors_replicas_______________1:1_______________2:2_______________3:3
    1m0s   1824.8   1847.5   17.3ms        0     6111         2037/13/0       2037/1981/6         2037/14/0
    2m0s   1555.2   1807.0   17.7ms        0     6174         2058/39/0      2058/1949/60         2058/39/0
    3m0s   1696.7   1736.8   18.4ms        0     6189         2063/67/0     2063/1904/111         2063/66/0
    4m0s   1692.2   1706.7   18.7ms        0     6234         2078/94/0     2078/1846/166         2078/96/0
    5m0s   1615.5   1686.3   19.0ms        0     6498        2166/127/0     2166/1856/219        2166/129/0
```

However, now that I've run all these final experiments, I'm realizing there's a big question I should have asked sooner about the latency numbers -- how could writes commit in an average of 38ms (or less) if the round-trip latency between all nodes is 50ms? I'll check into what's going on with that.

I'll also continue testing out some different things (particularly multiple nodes per locality and localities different distances apart) to make sure they behave reasonably and should add a unit test, but the code is ready for review.

@BramGruneir @petermattis

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/13797)
<!-- Reviewable:end -->

a-robinson committed:
storage: Transfer leases based on request load and latency

i.e. "Leaseholders follow the sun"

tamird commented:
Reviewed 9 of 9 files at r1.
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---

*[pkg/rpc/clock_offset.go, line 101 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13797#-KdpH75f7tihZgGKjXWa:-KdpH75g9zfGd_eZKk2w:bumxemp) ([raw file](https://github.com/cockroachdb/cockroach/blob/c45c1214c568027b372c1fbf3d3d42ffb9e4b92c/pkg/rpc/clock_offset.go#L101)):*
> ```Go
> // Latency returns the exponentially weighted moving average latency to the
> // given node address. Returns 0 if we do not yet have enough samples to return
> // a reliable average.
> ```

might be wise to return a Duration, bool pair

---

*[pkg/storage/allocator.go, line 502 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13797#-KdpHjdQl6SwyXH4vw_O:-KdpHjdQl6SwyXH4vw_P:b-mwbslx) ([raw file](https://github.com/cockroachdb/cockroach/blob/c45c1214c568027b372c1fbf3d3d42ffb9e4b92c/pkg/storage/allocator.go#L502)):*
> ```Go
> 	case decideWithoutStats:
> 	default:
> 		log.Errorf(ctx, "unexpected transfer decision %d", transferDec)
> ```

should these be Fatal? there's another above.

---

*[pkg/storage/allocator.go, line 533 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13797#-KdpIEsMRrhCiCf4KQRy:-KdpIEsMRrhCiCf4KQRz:bf8a247) ([raw file](https://github.com/cockroachdb/cockroach/blob/c45c1214c568027b372c1fbf3d3d42ffb9e4b92c/pkg/storage/allocator.go#L533)):*
> ```Go
> 	// On the other hand, if we don't have any stats with associated localities,
> 	// then do fall back to the algorithm that doesn't use request stats.
> 	delete(requestCounts, "")
> ```

is the empty key ever considered? perhaps this logic should be on the recording side

also, i'm nervous about the lack of tests for things like this and the escape hatch above - both should be reasonably easy to unit test.

---

*[pkg/storage/allocator.go, line 541 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13797#-KdpJ8416e1gNl6wlWFN:-KdpJ8416e1gNl6wlWFO:b-dvy0ie) ([raw file](https://github.com/cockroachdb/cockroach/blob/c45c1214c568027b372c1fbf3d3d42ffb9e4b92c/pkg/storage/allocator.go#L541)):*
> ```Go
> 	replicaLocalities := a.storePool.getLocalities(existing)
> 	for requestLocalityStr, count := range requestCounts {
> 		if requestLocalityStr == "" {
> ```

impossible; you deleted the empty key above, right?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13797)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
There should be an environment variable to disable the stats-based rebalancing completely in case it starts going nuts. 

---

Reviewed 9 of 9 files at r1.
Review status: all files reviewed at latest revision, 6 unresolved discussions.

---

*[pkg/storage/allocator.go, line 62 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13797#-KdxWZrvqYvIHwUsh_i9:-KdxWZrvqYvIHwUsh_iA:b-14hv0v) ([raw file](https://github.com/cockroachdb/cockroach/blob/c45c1214c568027b372c1fbf3d3d42ffb9e4b92c/pkg/storage/allocator.go#L62)):*
> ```Go
> 	// stats should be cleared out to avoid old data from overpowering new data.
> 	// Made configurable for the sake of testing.
> 	ResetLeaseTransferStatsPeriod = 10 * time.Minute
> ```

So with these values we'll be unable to do stat-based rebalancing 10% of the time, right? 

---

*[pkg/storage/allocator.go, line 474 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13797#-KdxYKkOJXHlTU4p1pjT:-KdxYKkPJyYD8u4SHQWH:b-q3p9fq) ([raw file](https://github.com/cockroachdb/cockroach/blob/c45c1214c568027b372c1fbf3d3d42ffb9e4b92c/pkg/storage/allocator.go#L474)):*
> ```Go
> var EnableLeaseRebalancing = envutil.EnvOrDefaultBool("COCKROACH_ENABLE_LEASE_REBALANCING", true)
> 
> type transferDecision int
> ```

Move this up to the top of the file (or at least before its first use)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13797#-:-KdxZwyQzjunjxc3VTrc:b5ts2uz)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Agreed that this needs an env var to disable and it should probably default to disabled until we have enabled for a while on test clusters.

---

Review status: all files reviewed at latest revision, 12 unresolved discussions.

---

*[pkg/storage/allocator.go, line 405 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13797#-Ke-4iX9bWtxOlTcGzhH:-Ke-4iX9bWtxOlTcGzhI:bt9gz1s) ([raw file](https://github.com/cockroachdb/cockroach/blob/c45c1214c568027b372c1fbf3d3d42ffb9e4b92c/pkg/storage/allocator.go#L405)):*
> ```Go
> 	}
> 
> 	transferDec, repl := a.shouldTransferLeaseUsingStats(ctx, sl, source, stats, existing)
> ```

Strange that you're always calling `shouldTransferLeaseUsingStats` here, especially given the above call to `shouldTransferLease` and that the transfer decision is only used if `checkTransferLeaseSource` is true. I'm not following why this call is necessary.

---

*[pkg/storage/allocator.go, line 508 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13797#-Ke-2ZKYc13XkUQG0hXR:-Ke-2ZKZP-tLiQZ6c4jh:b-4psxhl) ([raw file](https://github.com/cockroachdb/cockroach/blob/c45c1214c568027b372c1fbf3d3d42ffb9e4b92c/pkg/storage/allocator.go#L508)):*
> ```Go
> }
> 
> func (a Allocator) shouldTransferLeaseUsingStats(
> ```

The naming of this method is a bit confusing given that `shouldTransferLease` also takes the stats and, in fact, calls this method.

---

*[pkg/storage/allocator.go, line 521 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13797#-Ke-2ski7bBc8h3ZiGlE:-Ke-2ski7bBc8h3ZiGlF:b-c6b3p6) ([raw file](https://github.com/cockroachdb/cockroach/blob/c45c1214c568027b372c1fbf3d3d42ffb9e4b92c/pkg/storage/allocator.go#L521)):*
> ```Go
> 	requestCounts, requestCountsDur := stats.getRequestCounts()
> 	if requestCountsDur > ResetLeaseTransferStatsPeriod {
> 		stats.resetRequestCounts()
> ```

Wasn't the plan to move to exponential weighted decaying stats? Completing resetting the stats every 10m feels yucky.

---

*[pkg/storage/allocator.go, line 582 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13797#-Ke-3B10CIXRYS7_fJeV:-Ke-3B10CIXRYS7_fJeW:bya5jt1) ([raw file](https://github.com/cockroachdb/cockroach/blob/c45c1214c568027b372c1fbf3d3d42ffb9e4b92c/pkg/storage/allocator.go#L582)):*
> ```Go
> 
> 		// This is a bit of magic that was determined to work well via a bunch of
> 		// testing. See #13232 for context.
> ```

I think some of the thinking behind this magic should be commented here. I realize this is a work in progress, so fine if this is a `TODO(alex): document this magic`.

---

*[pkg/storage/allocator.go, line 606 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13797#-Ke-3VViqada-2rv6bmm:-Ke-3VViqada-2rv6bmn:b-wah0g8) ([raw file](https://github.com/cockroachdb/cockroach/blob/c45c1214c568027b372c1fbf3d3d42ffb9e4b92c/pkg/storage/allocator.go#L606)):*
> ```Go
> 				source.Capacity.LeaseCount, overfullLeaseThreshold, storeDesc.Capacity.LeaseCount,
> 				underfullLeaseThreshold, totalScore,
> 			)
> ```

Btw, something you've probably seen while extending this code: debugging allocator decisions can be a bit painful. I dream of a `/debug/allocator` page that can give us better insights, though I'm not sure what it would contain yet. Or perhaps the `/debug/range` page could provide details on what decision the allocator would make for every replica in the specified range. 

Cc @bramgruneir 

---

*[pkg/storage/allocator.go, line 616 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13797#-Ke-59ndEBYS-dXG1bR7:-Ke-59ndEBYS-dXG1bR8:b-ay0w2r) ([raw file](https://github.com/cockroachdb/cockroach/blob/c45c1214c568027b372c1fbf3d3d42ffb9e4b92c/pkg/storage/allocator.go#L616)):*
> ```Go
> 		return shouldTransfer, bestRepl
> 	}
> 	return shouldNotTransfer, bestRepl
> ```

Lots of logic here that feels unit testable, even if the heuristics themselves are tricky to test robustly. Perhaps parts of this method could be broken out. For example, computing the replica score given weights and latencies.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13797#-:-Ke-3zygG8rJiAU9vBg4:b-uwlksx)*
<!-- Sent from Reviewable.io -->

a-robinson commented:
Still need to work on unit tests and avoiding complete resets of replica stats. All other comments have been addressed.

---

Review status: all files reviewed at latest revision, 12 unresolved discussions.

---

*[pkg/rpc/clock_offset.go, line 101 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13797#-KdpH75f7tihZgGKjXWa:-Ke-K4EH3AX8W6-pmdbw:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/c45c1214c568027b372c1fbf3d3d42ffb9e4b92c/pkg/rpc/clock_offset.go#L101)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

might be wise to return a Duration, bool pair
</blockquote></details>

Done.

---

*[pkg/storage/allocator.go, line 62 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13797#-KdxWZrvqYvIHwUsh_i9:-Ke-yGXKGZMBOHfytQET:b17mgat) ([raw file](https://github.com/cockroachdb/cockroach/blob/c45c1214c568027b372c1fbf3d3d42ffb9e4b92c/pkg/storage/allocator.go#L62)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

So with these values we'll be unable to do stat-based rebalancing 10% of the time, right? 
</blockquote></details>

That would be the case, yes. I'm looking at switching to decaying rather than resetting the existing values, though.

---

*[pkg/storage/allocator.go, line 405 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13797#-Ke-4iX9bWtxOlTcGzhH:-Ke-OW_ndMujwY9cxOsX:b9fqd8v) ([raw file](https://github.com/cockroachdb/cockroach/blob/c45c1214c568027b372c1fbf3d3d42ffb9e4b92c/pkg/storage/allocator.go#L405)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Strange that you're always calling `shouldTransferLeaseUsingStats` here, especially given the above call to `shouldTransferLease` and that the transfer decision is only used if `checkTransferLeaseSource` is true. I'm not following why this call is necessary.
</blockquote></details>

Good catch, I had intended to replace the old check with this one. Removed it now.

Does the new comment clarify things at all?

---

*[pkg/storage/allocator.go, line 474 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13797#-KdxYKkOJXHlTU4p1pjT:-Ke-KTpbu9T8DxZj85JH:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/c45c1214c568027b372c1fbf3d3d42ffb9e4b92c/pkg/storage/allocator.go#L474)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Move this up to the top of the file (or at least before its first use)
</blockquote></details>

Done.

---

*[pkg/storage/allocator.go, line 502 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13797#-KdpHjdQl6SwyXH4vw_O:-Ke-KgCjgSF3CvYU0dNs:b10z35c) ([raw file](https://github.com/cockroachdb/cockroach/blob/c45c1214c568027b372c1fbf3d3d42ffb9e4b92c/pkg/storage/allocator.go#L502)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

should these be Fatal? there's another above.
</blockquote></details>

I guess? Done.

---

*[pkg/storage/allocator.go, line 508 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13797#-Ke-2ZKYc13XkUQG0hXR:-Ke-j6k8syYo_m2nVn-u:b-qllb8u) ([raw file](https://github.com/cockroachdb/cockroach/blob/c45c1214c568027b372c1fbf3d3d42ffb9e4b92c/pkg/storage/allocator.go#L508)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

The naming of this method is a bit confusing given that `shouldTransferLease` also takes the stats and, in fact, calls this method.
</blockquote></details>

Yeah, it is a bit funky. There was no reason for `shouldTransferLease` to be its own method, so I've just folded it into `ShouldTransferLease`.

---

*[pkg/storage/allocator.go, line 533 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13797#-KdpIEsMRrhCiCf4KQRy:-Ke-lUJ3-aIVNc7OkJuh:b-xv89kl) ([raw file](https://github.com/cockroachdb/cockroach/blob/c45c1214c568027b372c1fbf3d3d42ffb9e4b92c/pkg/storage/allocator.go#L533)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

is the empty key ever considered? perhaps this logic should be on the recording side

also, i'm nervous about the lack of tests for things like this and the escape hatch above - both should be reasonably easy to unit test.
</blockquote></details>

I'll send a separate PR to cut down on the number of requests without a source locality. We're not currently using this info for anything, but I've been operating under the assumption that there'll be some value in knowing how many requests are coming from nodes without locality info, since poorly configured localities is something we need to alert the user about and be aware of when debugging issues.

---

*[pkg/storage/allocator.go, line 541 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13797#-KdpJ8416e1gNl6wlWFN:-Ke-m2PwlP14xlZ1B5on:ba4ail5) ([raw file](https://github.com/cockroachdb/cockroach/blob/c45c1214c568027b372c1fbf3d3d42ffb9e4b92c/pkg/storage/allocator.go#L541)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

impossible; you deleted the empty key above, right?
</blockquote></details>

Yup, forgot to remove this after adding the above `delete` in order to make the `== 0` check more understandable.

---

*[pkg/storage/allocator.go, line 582 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13797#-Ke-3B10CIXRYS7_fJeV:-Ke-mCqnKbW-oOY9mgqF:b-l7sq4m) ([raw file](https://github.com/cockroachdb/cockroach/blob/c45c1214c568027b372c1fbf3d3d42ffb9e4b92c/pkg/storage/allocator.go#L582)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

I think some of the thinking behind this magic should be commented here. I realize this is a work in progress, so fine if this is a `TODO(alex): document this magic`.
</blockquote></details>

Yeah, good point. Put in the TODO.

---

*[pkg/storage/allocator.go, line 606 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13797#-Ke-3VViqada-2rv6bmm:-Ke-Let1X0rgdajr65YY:b-oxx6el) ([raw file](https://github.com/cockroachdb/cockroach/blob/c45c1214c568027b372c1fbf3d3d42ffb9e4b92c/pkg/storage/allocator.go#L606)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Btw, something you've probably seen while extending this code: debugging allocator decisions can be a bit painful. I dream of a `/debug/allocator` page that can give us better insights, though I'm not sure what it would contain yet. Or perhaps the `/debug/range` page could provide details on what decision the allocator would make for every replica in the specified range. 

Cc @bramgruneir 
</blockquote></details>

@bramgruneir and I should talk this week. It seems like it'd be perfectly feasible to expose the information being used to make decisions

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13797#-:-Ke-yv7MihK7u4cT5bws:b7eicfa)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 5 of 5 files at r2.
Review status: all files reviewed at latest revision, 8 unresolved discussions, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13797)*
<!-- Sent from Reviewable.io -->

a-robinson committed:
storage: Use a windowed decay algorithm in replicaStats

Part of me dislikes checking a clock on every request, but it's
probably a non-issue.

a-robinson commented:
Review status: 0 of 11 files reviewed at latest revision, 8 unresolved discussions, some commit checks pending.

---

*[pkg/storage/allocator.go, line 521 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13797#-Ke-2ski7bBc8h3ZiGlE:-Ke4I_osy7Me6unBbdib:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/c45c1214c568027b372c1fbf3d3d42ffb9e4b92c/pkg/storage/allocator.go#L521)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Wasn't the plan to move to exponential weighted decaying stats? Completing resetting the stats every 10m feels yucky.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13797)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 2 of 11 files at r3, 9 of 9 files at r4.
Review status: all files reviewed at latest revision, 15 unresolved discussions, some commit checks failed.

---

*[pkg/storage/replica_stats.go, line 28 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/13797#-Ke4eb6J0RvF-2KSqPip:-Ke4eb6Kh8EcIbRwqiiS:b-pj05g6) ([raw file](https://github.com/cockroachdb/cockroach/blob/e555aaf560b59ffbfe540d47cdad86d5a883112e/pkg/storage/replica_stats.go#L28)):*
> ```Go
> const (
> 	numWindows     = 5
> 	rotateInterval = 5 * time.Minute
> ```

i think you may want to give this one a type to prevent it from being used as a simple integer

---

*[pkg/storage/replica_stats.go, line 49 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/13797#-Ke4fUX45ToAF4RqP7Ty:-Ke4fUX45ToAF4RqP7Tz:bt081qc) ([raw file](https://github.com/cockroachdb/cockroach/blob/e555aaf560b59ffbfe540d47cdad86d5a883112e/pkg/storage/replica_stats.go#L49)):*
> ```Go
> 		syncutil.Mutex
> 		idx        int
> 		requests   []perLocalityCounts
> ```

can you document that this is a circular buffer? also, `numWindows` is a constant, so this can be an array rather than a slice.

you should just inline `numWindows` into the array definition, i don't think you'll need it anywhere else.

---

*[pkg/storage/replica_stats.go, line 80 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/13797#-Ke4gCtu0LbjP4l5tLLZ:-Ke4gCtvPjAhy8i5OJro:b-ocmb2x) ([raw file](https://github.com/cockroachdb/cockroach/blob/e555aaf560b59ffbfe540d47cdad86d5a883112e/pkg/storage/replica_stats.go#L80)):*
> ```Go
> 
> func (rs *replicaStats) rotateLocked() {
> 	rs.mu.idx = (rs.mu.idx + 1) % numWindows
> ```

s/numWindows/len(rs.mu.requests)/

---

*[pkg/storage/replica_stats.go, line 86 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/13797#-Ke4fqG2ZA0yzLlcp8s0:-Ke4fqG2ZA0yzLlcp8s1:b4kxm6g) ([raw file](https://github.com/cockroachdb/cockroach/blob/e555aaf560b59ffbfe540d47cdad86d5a883112e/pkg/storage/replica_stats.go#L86)):*
> ```Go
> // getRequestCounts returns the current per-locality request counts and the
> // amount of time over which the counts were accumulated.
> func (rs *replicaStats) getRequestCounts() (perLocalityCounts, time.Duration) {
> ```

this code needs tests

---

*[pkg/storage/replica_stats.go, line 94 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/13797#-Ke4gH6QgHTRNeSnNPpO:-Ke4gH6QgHTRNeSnNPpP:bysgsoc) ([raw file](https://github.com/cockroachdb/cockroach/blob/e555aaf560b59ffbfe540d47cdad86d5a883112e/pkg/storage/replica_stats.go#L94)):*
> ```Go
> 	counts := make(perLocalityCounts)
> 
> 	for i := 0; i < numWindows; i++ {
> ```

`for i := range rs.mu.requests`

---

*[pkg/storage/replica_stats.go, line 109 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/13797#-Ke4fyjJh6qlq7xU1oeh:-Ke4fyjJh6qlq7xU1oei:b29uoel) ([raw file](https://github.com/cockroachdb/cockroach/blob/e555aaf560b59ffbfe540d47cdad86d5a883112e/pkg/storage/replica_stats.go#L109)):*
> ```Go
> 	defer rs.mu.Unlock()
> 
> 	for i := 0; i < numWindows; i++ {
> ```

slight preference for `for i := range rs.mu.requests`

---

*[pkg/storage/replica_stats_test.go, line 26 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/13797#-Ke4eQTeNfXgdzDj3qpt:-Ke4eQTeNfXgdzDj3qpu:b-2w2uym) ([raw file](https://github.com/cockroachdb/cockroach/blob/e555aaf560b59ffbfe540d47cdad86d5a883112e/pkg/storage/replica_stats_test.go#L26)):*
> ```Go
> )
> 
> // TODO(DONOTMERGE): Add tests for windowing/decay logic.
> ```

commenting so this doesn't get lost

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13797)*
<!-- Sent from Reviewable.io -->

a-robinson commented:
Review status: all files reviewed at latest revision, 15 unresolved discussions, some commit checks failed.

---

*[pkg/storage/replica_stats.go, line 28 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/13797#-Ke4eb6J0RvF-2KSqPip:-Ke50DFW2mELKDzowFok:b5zsu8o) ([raw file](https://github.com/cockroachdb/cockroach/blob/e555aaf560b59ffbfe540d47cdad86d5a883112e/pkg/storage/replica_stats.go#L28)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

i think you may want to give this one a type to prevent it from being used as a simple integer
</blockquote></details>

It gets a type by virtue of time.Minute having a type. I was unable to pass this directly to a function that takes an int64 when writing the tests for this code.

---

*[pkg/storage/replica_stats.go, line 49 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/13797#-Ke4fUX45ToAF4RqP7Ty:-Ke51JNiBaRrvSDShUuQ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e555aaf560b59ffbfe540d47cdad86d5a883112e/pkg/storage/replica_stats.go#L49)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

can you document that this is a circular buffer? also, `numWindows` is a constant, so this can be an array rather than a slice.

you should just inline `numWindows` into the array definition, i don't think you'll need it anywhere else.
</blockquote></details>

Done.

---

*[pkg/storage/replica_stats.go, line 80 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/13797#-Ke4gCtu0LbjP4l5tLLZ:-Ke51k6qHqs3khLDgZiX:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e555aaf560b59ffbfe540d47cdad86d5a883112e/pkg/storage/replica_stats.go#L80)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

s/numWindows/len(rs.mu.requests)/
</blockquote></details>

Done.

---

*[pkg/storage/replica_stats.go, line 86 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/13797#-Ke4fqG2ZA0yzLlcp8s0:-Ke51iBixnzc83mU4sxF:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e555aaf560b59ffbfe540d47cdad86d5a883112e/pkg/storage/replica_stats.go#L86)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

this code needs tests
</blockquote></details>

Done.

---

*[pkg/storage/replica_stats.go, line 94 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/13797#-Ke4gH6QgHTRNeSnNPpO:-Ke51hwf2Z_51YPWbQ0t:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e555aaf560b59ffbfe540d47cdad86d5a883112e/pkg/storage/replica_stats.go#L94)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

`for i := range rs.mu.requests`
</blockquote></details>

Done.

---

*[pkg/storage/replica_stats.go, line 109 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/13797#-Ke4fyjJh6qlq7xU1oeh:-Ke51lYTjT4Ruy5ZXCoS:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e555aaf560b59ffbfe540d47cdad86d5a883112e/pkg/storage/replica_stats.go#L109)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

slight preference for `for i := range rs.mu.requests`
</blockquote></details>

Done.

---

*[pkg/storage/replica_stats_test.go, line 26 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/13797#-Ke4eQTeNfXgdzDj3qpt:-Ke527EnyfVr4Rqqd8N0:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e555aaf560b59ffbfe540d47cdad86d5a883112e/pkg/storage/replica_stats_test.go#L26)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

commenting so this doesn't get lost
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13797)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
Reviewed 3 of 12 files at r5, 9 of 9 files at r6.
Review status: all files reviewed at latest revision, 13 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13797)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 3 of 12 files at r5, 9 of 9 files at r6.
Review status: all files reviewed at latest revision, 8 unresolved discussions, some commit checks pending.

---

*[pkg/storage/replica_stats.go, line 28 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/13797#-Ke4eb6J0RvF-2KSqPip:-Ke56pkJ6U6Fzo9hXFVJ:bgx6jgg) ([raw file](https://github.com/cockroachdb/cockroach/blob/e555aaf560b59ffbfe540d47cdad86d5a883112e/pkg/storage/replica_stats.go#L28)):*
<details><summary><i>Previously, a-robinson (Alex Robinson) wrote…</i></summary><blockquote>

It gets a type by virtue of time.Minute having a type. I was unable to pass this directly to a function that takes an int64 when writing the tests for this code.
</blockquote></details>

Excellent, thanks.

---

*[pkg/storage/replica_stats.go, line 82 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/13797#-Ke57NDFonGSYhCaYDb9:-Ke57NDFonGSYhCaYDbA:bgsad8o) ([raw file](https://github.com/cockroachdb/cockroach/blob/1d5904e2a6462533022f6ac166298c6d54263338/pkg/storage/replica_stats.go#L82)):*
> ```Go
> }
> 
> func (rs *replicaStats) rotateIfNeededLocked(now time.Time) {
> ```

nit (if you eat more CI here): we usually name these `maybeDoSomething`

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13797)*
<!-- Sent from Reviewable.io -->

a-robinson commented:
Test for TransferLeaseTarget is up. More focused test on the details of the scoring will be up next.

---

Review status: 2 of 12 files reviewed at latest revision, 7 unresolved discussions.

---

*[pkg/storage/replica_stats.go, line 82 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/13797#-Ke57NDFonGSYhCaYDb9:-Ke582x9Zi3Nq4mAj8PR:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/1d5904e2a6462533022f6ac166298c6d54263338/pkg/storage/replica_stats.go#L82)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

nit (if you eat more CI here): we usually name these `maybeDoSomething`
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13797#-:-Ke5kVfNqDF7cpJXj93-:b-3ga6bn)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 1 of 10 files at r7, 9 of 9 files at r8.
Review status: all files reviewed at latest revision, 9 unresolved discussions, some commit checks pending.

---

*[pkg/storage/allocator_test.go, line 183 at r8](https://reviewable.io:443/reviews/cockroachdb/cockroach/13797#-Ke5lWNqy0sf1xIUgcfg:-Ke5lWNqy0sf1xIUgcfh:b-qkjpxh) ([raw file](https://github.com/cockroachdb/cockroach/blob/18f40e5a07c7f67fca8ed53849d05017e46f5c0d/pkg/storage/allocator_test.go#L183)):*
> ```Go
> 	stopper, g, manual, storePool, _ := createTestStorePool(
> 		TestTimeUntilStoreDeadOff, deterministic, true /* defaultNodeLiveness */)
> 	a := MakeAllocator(storePool, func(_ string) (time.Duration, bool) {
> ```

underscore not needed here (because all the arguments are unbound). there's another below

---

*[pkg/storage/allocator_test.go, line 1055 at r8](https://reviewable.io:443/reviews/cockroachdb/cockroach/13797#-Ke5mDc2B3F3McZqnU7_:-Ke5mDc2B3F3McZqnU7a:b-jdghim) ([raw file](https://github.com/cockroachdb/cockroach/blob/18f40e5a07c7f67fca8ed53849d05017e46f5c0d/pkg/storage/allocator_test.go#L1055)):*
> ```Go
> 		g.SetNodeDescriptor(&store.Node)
> 	}
> 	evenlyBalanced := newReplicaStats(clock, func(nodeID roachpb.NodeID) string {
> ```

this may be clearer as
```
localityFn := func(nodeID roachpb.NodeID) string {
  return localities[nodeID]
}
evenlyBalanced := func() *replicaStats {
  rs := newReplicaStats(clock, localityFn)
  rs.record(1)
  rs.record(2)
  rs.record(3)
  return rs
}()
```
and so on for the other cases below.

---

*[pkg/storage/allocator_test.go, line 1095 at r8](https://reviewable.io:443/reviews/cockroachdb/cockroach/13797#-Ke5nKLCio1qPyx0WI99:-Ke5nKLCio1qPyx0WI9A:b-xt170b) ([raw file](https://github.com/cockroachdb/cockroach/blob/18f40e5a07c7f67fca8ed53849d05017e46f5c0d/pkg/storage/allocator_test.go#L1095)):*
> ```Go
> 	}
> 
> 	// TODO(peter): Add test cases for different latencies
> ```

should this have your name?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13797)*
<!-- Sent from Reviewable.io -->

a-robinson commented:
PTAL.

Also the original mystery of allocsim-reported latencies being lower than the configured RTT has been explained by the revelation that we open up raft streams in both directions and rarely receive anything on the outgoing streams, so the injected sleep isn't taking effect on most raft traffic. I don't yet have an obvious fix that'll work on MacOS while still allowing per-node latency distinctions. We could instead simplify the problem down to only supporting identical latency between all nodes on MacOS and only supporting different latencies between different nodes on Linux.

---

Review status: 10 of 12 files reviewed at latest revision, 9 unresolved discussions, some commit checks pending.

---

*[pkg/storage/allocator.go, line 616 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13797#-Ke-59ndEBYS-dXG1bR7:-Ke9PKR0xnOXASD2IPIy:b51dhys) ([raw file](https://github.com/cockroachdb/cockroach/blob/c45c1214c568027b372c1fbf3d3d42ffb9e4b92c/pkg/storage/allocator.go#L616)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Lots of logic here that feels unit testable, even if the heuristics themselves are tricky to test robustly. Perhaps parts of this method could be broken out. For example, computing the replica score given weights and latencies.
</blockquote></details>

Added some tests.

---

*[pkg/storage/allocator_test.go, line 183 at r8](https://reviewable.io:443/reviews/cockroachdb/cockroach/13797#-Ke5lWNqy0sf1xIUgcfg:-Ke9PCqQL0ZyDs_2LHbA:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/18f40e5a07c7f67fca8ed53849d05017e46f5c0d/pkg/storage/allocator_test.go#L183)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

underscore not needed here (because all the arguments are unbound). there's another below
</blockquote></details>

Done.

---

*[pkg/storage/allocator_test.go, line 1055 at r8](https://reviewable.io:443/reviews/cockroachdb/cockroach/13797#-Ke5mDc2B3F3McZqnU7_:-Ke9PEeG9HQer13aa2ML:b-75ypwt) ([raw file](https://github.com/cockroachdb/cockroach/blob/18f40e5a07c7f67fca8ed53849d05017e46f5c0d/pkg/storage/allocator_test.go#L1055)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

this may be clearer as
```
localityFn := func(nodeID roachpb.NodeID) string {
  return localities[nodeID]
}
evenlyBalanced := func() *replicaStats {
  rs := newReplicaStats(clock, localityFn)
  rs.record(1)
  rs.record(2)
  rs.record(3)
  return rs
}()
```
and so on for the other cases below.
</blockquote></details>

Cleaned up a bit and commented.

---

*[pkg/storage/allocator_test.go, line 1095 at r8](https://reviewable.io:443/reviews/cockroachdb/cockroach/13797#-Ke5nKLCio1qPyx0WI99:-Ke9PG09b_p1HXkgw6iR:b29k1bd) ([raw file](https://github.com/cockroachdb/cockroach/blob/18f40e5a07c7f67fca8ed53849d05017e46f5c0d/pkg/storage/allocator_test.go#L1095)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

should this have your name?
</blockquote></details>

It shouldn't exist at all.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13797#-:-Ke9Pz9_sSEqchpjjf0U:b-ow8ccp)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
Can we go back to injecting latency in SendMsg instead of RecvMsg? They were believed to be nearly equivalent in https://github.com/cockroachdb/cockroach/pull/13534#issuecomment-280097792 (and the next comment says that throughput increased when we switched from SendMsg to RecvMsg, which should have been a red flag in this context)

Do we need to inject latency on both Send and Recv? There are four "sides" here, client and server, send and receive. I think we need to inject on both Send and Recv since we're only doing it on the client. 

---

Reviewed 1 of 10 files at r7, 9 of 9 files at r8, 2 of 2 files at r9.
Review status: all files reviewed at latest revision, 9 unresolved discussions, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13797#-:-KeAb4XskEj7NLbfZM05:b-337gqr)*
<!-- Sent from Reviewable.io -->

a-robinson commented:
It's not quite equivalent to real network latency, but perhaps it's close enough to work for testing purposes?

Using 25ms connection latency (i.e. 50ms round trip) for all traffic on my computer, I get:
```
    1m0s    203.0    168.4  189.7ms        0       69            23/4/9           23/14/1            23/2/2
    2m0s    190.9    182.5  175.2ms        0      117           39/6/10           39/24/5            39/6/3
    3m0s    186.0    185.9  172.1ms        0      177           59/9/10          59/37/10            59/7/7
    4m0s    194.7    186.4  171.6ms        0      216          72/12/15          72/45/23          72/12/16
    5m0s    184.4    186.9  171.1ms        0      219          73/12/20          73/47/29          73/11/19
    6m0s    184.8    187.4  170.7ms        0      360         120/21/22         120/73/41         120/20/22
    7m0s    191.7    187.3  170.8ms        0      411         137/25/28         137/87/55         137/23/27
    8m0s    192.8    187.4  170.7ms        0      414         138/24/34         138/89/62         138/23/30
    9m0s    190.1    187.3  170.8ms        0      414         138/23/42         138/91/69         138/22/32
   10m1s    187.9    187.3  170.8ms        0      417         139/23/44         139/91/74         139/23/33
   11m0s    190.7    187.5  170.7ms        0      510         170/27/47        170/109/83         170/27/35
   12m0s    187.4    187.5  170.7ms        0      690         230/42/52       230/141/105         230/41/40
```

Using a 50ms sleep in `SendMsg`, I get noticeably worse (but not insanely worse) performance:
```
    1m0s    207.3    170.9  187.0ms        0       69            23/5/2           23/16/3            23/2/1
    2m0s    170.5    167.3  191.1ms        0      117            39/6/4          39/26/10            39/7/2
    3m0s    169.2    165.0  193.8ms        0      135            45/6/8          45/30/14            45/9/4
    4m0s    150.5    164.3  194.7ms        0      216          72/12/13          72/47/30           72/12/9
    5m0s    148.0    163.2  196.0ms        0      219          73/12/20          73/49/37          73/12/12
    6m0s    140.9    162.0  197.5ms        0      240          80/12/26          80/54/44          80/14/13
```

Using a 25ms sleep in both `SendMsg` and `RecvMsg`, I get way better performance (that kept getting better, so I stopped early):
```
1m0s    296.9    235.6  135.7ms        0       69            23/4/3           23/15/3           23/4/12
```

So the `SendMsg` 50ms sleep is closest, but still not correct. Using entire machine latency is good enough for the tests that I've been running, so perhaps it is fine to just leave the fancy stuff to Linux.

---

Review status: all files reviewed at latest revision, 9 unresolved discussions, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13797#-:-KeB4EkZUNkme-HStaUH:bgswwj7)*
<!-- Sent from Reviewable.io -->

petermattis commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: all files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13797#-:-KeZomumyGlw_viYBWUt:bnfp4nl)*
<!-- Sent from Reviewable.io -->

a-robinson commented:
@BramGruneir, if you have comments can you leave them on here this afternoon? I'm still playing with the specifics of the formula on some different cluster configurations, but would like to get this checked in soon.

BramGruneir commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 1 of 9 files at r1, 3 of 5 files at r2, 1 of 12 files at r5, 1 of 10 files at r7, 4 of 9 files at r8, 1 of 10 files at r10, 9 of 9 files at r11.
Review status: all files reviewed at latest revision, 4 unresolved discussions, all commit checks successful.

---

*[pkg/storage/allocator.go, line 559 at r11](https://reviewable.io:443/reviews/cockroachdb/cockroach/13797#-Ke_SdDUBsaiKMJa9hiT:-Ke_SdDVdD89kz6xGovF:b-cass4h) ([raw file](https://github.com/cockroachdb/cockroach/blob/92f27c51a732aa4a5b54eceac3da0c484a3023a3/pkg/storage/allocator.go#L559)):*
> ```Go
> 			continue
> 		}
> 		storeDesc, ok := a.storePool.getStoreDescriptor(repl.StoreID)
> ```

Consider adding a getStoreListForIDs that takes a list of storeIDs, so we don't lock and unlock the store for each existing replica.  

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13797#-:-Ke_TmcRoCFUa81lJAV4:bnfp4nl)*
<!-- Sent from Reviewable.io -->

a-robinson commented:
Review status: 11 of 12 files reviewed at latest revision, 4 unresolved discussions, all commit checks successful.

---

*[pkg/storage/allocator.go, line 559 at r11](https://reviewable.io:443/reviews/cockroachdb/cockroach/13797#-Ke_SdDUBsaiKMJa9hiT:-KedM_Rxe0Y81pfC8SGU:bp30rgb) ([raw file](https://github.com/cockroachdb/cockroach/blob/92f27c51a732aa4a5b54eceac3da0c484a3023a3/pkg/storage/allocator.go#L559)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary><blockquote>

Consider adding a getStoreListForIDs that takes a list of storeIDs, so we don't lock and unlock the store for each existing replica.  
</blockquote></details>

I'm going to skip mucking the code up with that for now, but will keep it in mind if profiles show much contention on the store pool's lock.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13797)*
<!-- Sent from Reviewable.io -->
